tfutils/base.py

```diff
@@ -816,49 +950,112 @@ def train(sess,
         - dbinterface (DBInterface object)
             Saver through which to save results
+        - train_loop (callable withs args: sess and train_targets)
+            Callable that specifies a custom training loop
         - train_targets (dict of tensorflow nodes)
             Targets to train. One item in this dict must be "optimizer" or similar
             to make anything happen
+        - num_minibatches (int)
+            How many minibatches to use to before applying gradient update.
         - num_steps (int)
             How many steps to train to before quitting
-        - valid_targets (dict of tensorflow objects, default: None)
+        - validdation_targets (dict of tensorflow objects, default: None)
             Objects on which validation will be computed
         - thres_loss (float, default: 100)
             If loss exceeds this during training, HiLossError is thrown
+
     """
+    # Collect args in a dict of lists
+    train_args = {
+        'sess': sess,
+        'queues': queues,
+        'num_steps': num_steps,
+        'thres_loss': thres_loss,
+        'train_loop': train_loop,
+        'global_step': global_step,
+        'dbinterface': dbinterface,
+        'train_targets': train_targets,
+        'validate_first': validate_first,
+        'num_minibatches': num_minibatches,
+        'validation_targets': validation_targets}
+
+    # Convert to a list of dicts
+    trargs = [{key: value[i] for (key, value) in train_args.items()}
+              for i in range(len(train_targets))]
+
+    num_steps = [t['num_steps'] for t in trargs]
+    steps = [t['global_step'].eval(session=t['sess']) for t in trargs]
+
+    # Start queues and initial validation
+    for (step, trarg) in zip(steps, trargs):
+
+        if step >= trarg['num_steps']:
+            log.info('Training cancelled since step ({}) is >= num_steps ({})'.
+                     format(step, trarg['num_steps']))
+            return
 
-    step = global_step.eval(session=sess)
+        log.info('Training beginning ...')
+        trarg['coord'], trarg['threads'] = start_queues(trarg['sess'])
+
+        if step == 0:
+            trarg['dbinterface'].start_time_step = time.time()
+            if trarg['validate_first']:
+                valid_res = run_targets_dict(trarg['sess'],
+                                             trarg['validation_targets'],
+                                             dbinterface=trarg['dbinterface'])
+    sess = train_args['sess'][0]
+    train_loop = train_args['train_loop'][0]
+    train_targets = train_args['train_targets']
+
+    # Run training
+    while any(step < num_step for (step, num_step) in zip(steps, num_steps)):
+
+        # Make a single call to sess.run to produce a list of results.
+        # TODO: call default train_loop which has minibatch option and is
+        # defined in base.py
+
+        start_time_step = time.time()
+        if train_loop is not None:
```

Comment:
there shouldn't be a case here.   there should be a default train_loop function, which is what is called unless the train_loop is specified.   moreover,  arguments to the function be allowed to be specified by the user